tests/unit/h/views/api/groups_test.py

```diff
@@ -8,477 +8,606 @@
+class TestGroups:
+    def test_it_without_request_params(
+        self, pyramid_request, group_list_service, GroupsJSONPresenter
+    ):
+        response = views.groups(pyramid_request)
 
         group_list_service.request_groups.assert_called_once_with(
-            user=None, authority=None, document_uri=None
+            user=pyramid_request.user,
+            authority=None,
+            document_uri=None,
+        )
+        GroupsJSONPresenter.assert_called_once_with(
+            group_list_service.request_groups.return_value, pyramid_request
         )
+        GroupsJSONPresenter.return_value.asdicts.assert_called_once_with(expand=[])
+        assert response == GroupsJSONPresenter.return_value.asdicts.return_value
+
+    def test_it_with_request_params(
+        self, pyramid_request, group_list_service, GroupsJSONPresenter
+    ):
+        pyramid_request.GET.add("expand", sentinel.expand_1)
+        pyramid_request.GET.add("expand", sentinel.expand_2)
+        pyramid_request.params["authority"] = sentinel.authority
+        pyramid_request.params["document_uri"] = sentinel.document_uri
 
-    def test_proxies_request_params(self, anonymous_request, group_list_service):
-        anonymous_request.params["document_uri"] = "http://example.com/thisthing.html"
-        anonymous_request.params["authority"] = "foo.com"
-        views.groups(anonymous_request)
+        views.groups(pyramid_request)
 
         group_list_service.request_groups.assert_called_once_with(
-            user=None,
-            authority="foo.com",
-            document_uri="http://example.com/thisthing.html",
+            user=pyramid_request.user,
+            authority=sentinel.authority,
+            document_uri=sentinel.document_uri,
+        )
+        GroupsJSONPresenter.return_value.asdicts.assert_called_once_with(
+            expand=[sentinel.expand_1, sentinel.expand_2]
         )
 
-    @pytest.mark.parametrize(
-        "expand",
-        ([], ["organization"], ["organization", "scopes"]),
-    )
-    def test_returns_dicts_from_presenter(
-        self,
-        anonymous_request,
-        open_groups,
-        group_list_service,
-        GroupsJSONPresenter,
-        expand,
-    ):
-        for param in expand:
-            anonymous_request.GET.add("expand", param)
 
-        group_list_service.request_groups.return_value = open_groups
+@pytest.mark.usefixtures("group_service", "group_create_service")
+class TestCreate:
+    def test_it_validates_the_request(self, pyramid_request, CreateGroupAPISchema):
+        views.create(pyramid_request)
 
-        result = views.groups(anonymous_request)
+        CreateGroupAPISchema.assert_called_once_with(
+            default_authority=pyramid_request.default_authority,
+            group_authority=sentinel.effective_authority,
+        )
+        CreateGroupAPISchema.return_value.validate.assert_called_once_with(
+            pyramid_request.json_body
+        )
```

Comment:
All the views need a `test_it_validates_the_request()` test like this one, plus each view also needs the `test_it_when_request_json_body_isnt_valid_json()` and `test_it_when_request_json_body_is_invalid()` tests below as well. It's a lot of repetitive tests, and the reason is that the each of the views contains repetitive code to validate the request:

```python
appstruct = CreateGroupAPISchema(
    default_authority=request.default_authority,
    group_authority=request.effective_authority,
).validate(_json_payload(request))
```

The way this should work is how permissions work in Pyramid, and how schema validation works in LMS - the view just specifices the schema in its `@view_config` decorator and then the already-validated request data is passed into the view as an argument or, if the request isn't valid, the view is never called at all. Then the views wouldn't need to repeat code like the above, and the tests for each view wouldn't need to test it, and things would be less error-prone.

```python
@view_config(..., schema=CreateGroupAPISchema)
```

For now I don't want to get into refactoring the code so I've gone with the repetitive unit tests. But at least all the validation-related tests are separated out into separate tests which will be easy to delete if we ever refactor the code and don't need them anymore.